genarg related cleanups (#19646)
coq/coq: doc/sphinx/language/core/assumptions.rst

the type of the variable as follows:
:n:`(@ident : @type := @term)`.

> `(x : T | P)` is syntactic sugar for `(x : Stdlib.Init.Specif.sig (fun x : T => P))`
> i.e. `(x : {x : T | P})` (it is implemented using the first, so changing the
> `{x : T | P}` notation will not change the meaning of `(x : T | P)`,
> and changing the implicit arguments of `sig` will break `(x : T | P)`).

Reviewer: ```suggestion
`(x : T | P)` is syntactic sugar for `(x : Stdlib.Init.Specif.sig (fun x : T => P))`
i.e. `(x : {x : T | P})`.  It is implemented using the first, so changing the
`{x : T | P}` notation will not change the meaning of `(x : T | P)`.
Changing the implicit arguments of `sig` will break `(x : T | P)`).
```

Can you say more about the "i.e. `(x : {x : T | P})`"?  You seem to rely to a non-obvious equivalence, would be better to explain just a little.

Reviewer 2: I like the original phrasing better than this suggestion.  I'd be fine with dropping the parenthetical and leaving it a single sentence, if that's what you're objecting to @jfehrle .  That is,
```suggestion
`(x : T | P)` is syntactic sugar for `(x : Stdlib.Init.Specif.sig (fun x : T => P))`
i.e. `(x : {x : T | P})`. It is implemented using `(x : Stdlib.Init.Specif.sig (fun x : T => P))`, so changing the
`{x : T | P}` notation will not change the meaning of `(x : T | P)`,
and changing the implicit arguments of `sig` will break `(x : T | P)`.
```
though I think this already might be slightly worse.

> Can you say more about the "i.e. (x : {x : T | P})"? You seem to rely to a non-obvious equivalence

Is the thing that you're missing that ` {x : T | P}` is a notation for `Stdlib.Init.Specif.sig (fun x : T => P)`?  (I think this is relatively obvious to anyone who uses either of these constructions frequently.)

Reviewer: ```suggestion
`(x : T | P)` is syntactic sugar for `(x : Stdlib.Init.Specif.sig (fun x : T => P))`.
`(x : {x : T | P})` is a notation for the same term.  Changing the notation
will not change the meaning of `(x : T | P)`, while
changing the implicit arguments of `sig` will break `(x : T | P)`).
```
I think this is better.  WDYT?

> relatively obvious to anyone who uses either of these constructions frequently.

We want the doc to be readily understood even by those who are new to the construction.

Reviewer 2: How about
```suggestion
`(x : T | P)` is syntactic sugar for `(x : Stdlib.Init.Specif.sig (fun x : T => P))`,
which would more typically be written `(x : {x : T | P})`.
Since `(x : T | P)` uses `sig` directly, 
changing the notation `{x : T | P}`
will not change the meaning of `(x : T | P)`, while
changing the implicit arguments of `sig` will break `(x : T | P)`).
```
?

Reviewer: "more typically written" meaning as typically written in the mathematical literature?  Is so, maybe add "in the literature" to be clear, since `(x : {x : T | P})` differs from the notation `{x : T | P}`.

But I'm curious about:
- Why does `(x : T | P)` exist in the grammar at all rather than as another notation?  Is it that notations with parentheses don't work?
- If users commonly want to redefine `{x : T | P}` (why?), then maybe replace the last sentence with "If you wish to redefine the `{x : T | P}` notation, you can use `(x : T | P)` to refer to the original definition.  Note that changing the implicit arguments of `sig` will break `(x : T | P)`."  
- How often will users want to use both the original definition and a changed notation?  Is it also common for users to want to redefine `sig`?  And why?

If you agree on (some of) these, maybe update your suggestion to make it easy for @SkySkimmer to adopt.

Author: >Why does (x : T | P) exist in the grammar at all rather than as another notation? Is it that notations with parentheses don't work?

notations are for terms, but this is a binder.

Reviewer 2: > "more typically written" meaning as typically written in the mathematical literature?

No, I mean in Coq developments.  I believe most users are not aware of `(x : T | P)` (I wasn't), and `(x : {x : T | P})` is the natural combination of `(x : A)` and `{x : T | P}`.

Your second and third suggestions/questions are not valid.  The text in the doc is a warning, not a suggestion.  Most users won't redefine the notation, but, e.g., HoTT and other foundations of math or alternate-stdlib developments might. In such cases, it's worth noting that this notation won't follow such a change.  Users are more likely to change the implicit argument status of `sig` (redefining it is impossible unless you pass `-noinit` / give your own stdlib).  And as @SkySkimmer points out, "If you wish to redefine the {x : T | P} notation, you can use (x : T | P) to refer to the original definition." is false.

Reviewer: Right.  But if it's a binder, then how are notations relevant here?

Reviewer: @JasonGross HoTT and these other developments redefine `{x : T | P}` or 1`sig` or both?  If so, maybe say something like "Some libraries, such as HoTT, may  a different definition of <whatever it is>.  In this case, be aware that the definition of `(x : T | P)` won't change.

Would it be useful/possible to have a way for the user to redefine what `(x : T | P)` generates?

Reviewer 2: > > notations are for terms, but this is a binder.
> 
> Right. But if it's a binder, then how are notations relevant here?

There's a grammar entry for binders, no? 

> @JasonGross HoTT and these other developments redefine {x : T | P} or 1sig or both?

Both, but the way it's done makes `(x : T | P)` invalid, I believe.

I maintain that the sentence or two we're bikeshedding over here is intended to be a short warning for advanced users and is not worth spending this much time on.  I believe it's more important to have the warning be accurate and precise than it is to make it accessible to first-time reference manual readers.

> Would it be useful/possible to have a way for the user to redefine what `(x : T | P)` generates?

Ideally, yes.  Currently, no.  I've left a comment to this effect at https://github.com/coq/coq/pull/19645/files#r1790790204.

Reviewer: It's worth the time to make it understandable.  It doesn't have to go into great detail.  And usually best to start from the user's perspective.  I'm not a first-time manual reader, either.

Reviewer 3: I happened to see this conversation and I agree with @JasonGross that this is not worth bikeshedding over. I don't like the argument that it's worth the time to make it understandable because there's finite time and it should be spent where valuable, and this is quite a niche part of the manual.

I also agree with Jason that "it is implemented using ..." is nice because it is technically accurate and clearly signals that this is for advanced users and a casual user can ignore the rest of the sentence, which allows to spend less time on its wording.

Author: This thread should be on https://github.com/coq/coq/pull/19645 not here
